biot_savart_integrators/generic/line_elements.py

```diff
@@ -0,0 +1,68 @@
+"""Numerical routines for Biot-Savart integration of a generic current curve
+
+The routines expect (dask-chunked) xarray inputs in order to facilitate
+memory-efficient and parallelized computation.
+"""
+import xarray as xr
+from .integrand import biot_savart_integrand as bsintegrand
+from .integrand import biot_savart_potential_integrand as bs_pot_integrands
+
+
+def biot_savart_integral(r: xr.DataArray, r_c: xr.DataArray,
+                         dl: xr.DataArray, j: xr.DataArray,
+                         integration_dim: str, spatial_dim: str) -> xr.DataArray:
+    """Numerical integration of B-S law for the set of line elements.
+
+    :param r: Positions where the magnetic field is evaluated.
+    :type r: xr.DataArray
+    :param r_c: Centers of line current elements.
+    :type r_c: xr.DataArray
+    :param dl: Lengt of the current element (scalar).
+    :type dl: xr.DataArray
+    :param j: Spatial current flowing through the element in Amps.
+    :type j: xr.DataArray
+    :param integration_dim: Dimension name over which index current
+                            elements.
+    :type integration_dim: str
+    :param spatial_dim: Name of the spetial dimension.
+    :type spatial_dim: str
+    """
+    j_int = j * dl
+    integrand = bsintegrand(r, r_c, j_int, spatial_dim)
+
+    # TODO: Trapez rule used by integrate is more precise method then simple
+    # summation. However, this method may brake if the coordinates are assign
+    # to the integration_dim
+    # integral = integrand.sum(integration_dim)
+    integral = integrand.integrate(integration_dim)
```

Comment:
This should be resolved before the merge. See the todo note.

Reply:
Ok, a simple solution would be to add an optional parameter `integral_method='trapz'` in a similar fashion as you do in pleque and add a warning about this into the docstring.

Reply:
I suppose that `method='sum'` would be something like `intergand.sum(integration_dim) * dl`?

Reply:
Nope. `integrate` here only do "summation" of elements of the `integrand`, but with Trapez rule (see the example below). `dl` element is already included in `j`.

https://github.com/smartass101/biot_savart_integrators/blob/32d294788bf788deba173627ff59e8ad782718e3/biot_savart_integrators/generic/curve.py#L12

If you assign some coordinates (for example `l = cumsum(dl)`) the integration will brake. 

```python
In [3]: a = xr.DataArray(np.linspace(1, 20, 20))                                                                                     

In [4]: a                                                                                                                            
Out[4]: 
<xarray.DataArray (dim_0: 20)>
array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12., 13.,
       14., 15., 16., 17., 18., 19., 20.])
Dimensions without coordinates: dim_0

In [5]: a.sum()                                                                                                                      
Out[5]: 
<xarray.DataArray ()>
array(210.)

In [8]: a.integrate("dim_0")                                                                                                         
Out[8]: 
<xarray.DataArray ()>
array(199.5)


```

Reply:
A possible solution would be something like this (untested;  adjusted your function):

```python
def biot_savart_integral(r, r_c, integration_dim, spatial_dim, I=1):
    # TODO make sure won't divide by coord - integration should cancel, but how accurately at edges?
    dl = r_c.differentiate(integration_dim)
    j = I*dl
    dl = (dl ** 2).sum(spatial_dim) ** 0.5 # normalize to be length element, not vector
    j = j / dl # remove length element

    integrand = bsintegrand(r, r_c, j, spatial_dim)
    integrand[integration_dim] = np.cumsum(dl)

    integral = integrand.integrate(integration_dim)
    return integral
```

or

```python
def biot_savart_integral(r, r_c, integration_dim, spatial_dim, I=1):
    # TODO make sure won't divide by coord - integration should cancel, but how accurately at edges?
    dl = r_c.differentiate(integration_dim)
    j = I*dl
    integrand = bsintegrand(r, r_c, j, spatial_dim)
    
    integrand[integration_dim] = np.cumsum(dl)
    integral = integrand.integrate(integration_dim) / dl.sum()
    return integral
```

Reply:
Did you mean `integrand.coords[integration_dim] = np.cumsum(dl)` ?

I suppose the simplest solution would be just to use `.sum()` instead of `.integrate()`. Perhaps the possible inaccuracy of `sum` is not  worth all these workarounds?